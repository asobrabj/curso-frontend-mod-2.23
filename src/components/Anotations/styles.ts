import styled from 'styled-components'

export const Container = styled.main`
  padding: 32px;
  margin: auto;
  z-index: 0;
`
export const ContainerCards = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  gap: 32px;

  @media (max-width: 820px) {
    grid-template-columns: 1fr 1fr;
  }

  @media (max-width: 600px) {
    display: block;
  }
`

export const ContainerArrows = styled.div`
  display: flex;
  justify-content: center;
  margin-top: 24px;

  > button {
    border: none;
    font-size: 23px;
    margin-right: 16px;
    cursor: pointer;

    &:active {
      font-size: 22px;
      &.mg {
        margin-right: 17px;
      }
    }

    &:disabled {
      cursor: auto;
      > svg {
        fill: #777;
      }
    }
  }
`
export const ContainerNotFound = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`
export const Icon = styled.i`
  font-size: 150px;
  color: ${(props) => props.theme.color.shadow};
`
export const Paragraph = styled.p`
  font-size: 32px;

  > span {
    font-style: italic;
    &:hover {
      cursor: pointer;
      text-decoration: underline;
    }
  }
`
