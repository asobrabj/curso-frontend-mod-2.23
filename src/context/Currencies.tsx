import { createContext, useContext, useState } from 'react';
import instance from '../service/axiosaInstace';

export interface ICurrencyContext {
  to: string;
  setTo: React.Dispatch<React.SetStateAction<string>>;
  from: string;
  setFrom: React.Dispatch<React.SetStateAction<string>>;
  amount: string;
  setAmount: React.Dispatch<React.SetStateAction<string>>;
  fecthConversion: (
    to: string,
    from: string,
    amount: string,
  ) => Promise<string>;
  error: boolean;
}

interface ICurrencyProvider {
  children: React.ReactNode;
}

const CurrencyContext = createContext<ICurrencyContext>({} as ICurrencyContext);

export const CurrencyProvider: React.FC<ICurrencyProvider> = ({ children }) => {
  const [error, setError] = useState<boolean>(false);
  const [to, setTo] = useState<string>('');
  const [from, setFrom] = useState<string>('');
  const [amount, setAmount] = useState<string>('');

  const fecthConversion = async (
    to: string,
    from: string,
    amount: string,
  ): Promise<string> => {
    try {
      const response = await instance.get(`${to}/${from}/${amount}`);
      return response.data.conversion_result;
    } catch (err) {
      setError(true);
      return 'Erro na convers√£o';
    }
  };

  return (
    <CurrencyContext.Provider
      value={{
        fecthConversion,
        error,
        amount,
        from,
        to,
        setTo,
        setAmount,
        setFrom,
      }}
    >
      {children}
    </CurrencyContext.Provider>
  );
};

export const useCurrency = () => {
  return useContext(CurrencyContext);
};
