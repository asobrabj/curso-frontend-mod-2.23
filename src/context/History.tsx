/* eslint-disable react-refresh/only-export-components */
import { createContext, useContext, useEffect, useState } from 'react';

export interface IDataCurrency {
  to: string;
  from: string;
  amount: string;
  result: string;
  date: string;
}

export interface IHistoryContext {
  setHistory: (data: IDataCurrency) => void;
  history?: Array<IDataCurrency>;
}

interface IHistoryProvider {
  children: React.ReactNode;
}

const HistoryContext = createContext<IHistoryContext>({} as IHistoryContext);

export const HistoryProvider: React.FC<IHistoryProvider> = ({ children }) => {
  const [history, setHistoryState] = useState<Array<IDataCurrency>>();

  useEffect(() => {
    const currentData = window.localStorage.getItem('history');
    const newData = currentData ? JSON.parse(currentData) : [];
    setHistoryState(newData);
  }, []);

  const setHistory = (data: IDataCurrency) => {
    const currentData = window.localStorage.getItem('history');
    const newData = currentData ? JSON.parse(currentData) : [];
    newData.push(data);
    setHistoryState(newData);
    window.localStorage.setItem('history', JSON.stringify(newData));
    return;
  };

  return (
    <HistoryContext.Provider value={{ setHistory, history }}>
      {children}
    </HistoryContext.Provider>
  );
};

export const useHistory = () => {
  return useContext(HistoryContext);
};
