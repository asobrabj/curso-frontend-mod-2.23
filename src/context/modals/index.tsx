import { createContext, useContext, useState } from 'react'
import IAnotation from '../../types/anotation'

export interface IModalContext {
  isVisible: boolean
  setIsVisible: React.Dispatch<React.SetStateAction<boolean>>
  isForm: boolean
  setIsForm: React.Dispatch<React.SetStateAction<boolean>>
  dataCard: IAnotation
  setDataCard: React.Dispatch<React.SetStateAction<IAnotation>>
  idCard: number | null
  setIdCard: React.Dispatch<React.SetStateAction<number | null>>
}

interface IModalProvider {
  children: React.ReactNode
}

const ModalContext = createContext<IModalContext>({} as IModalContext)

export function ModalProvider({ children }: IModalProvider) {
  const [isVisible, setIsVisible] = useState<boolean>(false)
  const [isForm, setIsForm] = useState<boolean>(false)
  const [idCard, setIdCard] = useState<number | null>(null)
  const [dataCard, setDataCard] = useState<IAnotation>({
    id: null,
    concluded: false,
    date: '',
    description: '',
    name: '',
    category: '',
    priority: '',
    time: '',
  })

  return (
    <ModalContext.Provider
      value={{
        idCard,
        setIdCard,
        isVisible,
        setIsVisible,
        isForm,
        setIsForm,
        dataCard,
        setDataCard,
      }}
    >
      {children}
    </ModalContext.Provider>
  )
}

export function useModal() {
  return useContext(ModalContext)
}
