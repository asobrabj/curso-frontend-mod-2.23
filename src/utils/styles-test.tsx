import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
