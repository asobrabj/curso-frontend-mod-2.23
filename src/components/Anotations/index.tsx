import React, { useEffect, useState } from 'react'
import { FaArrowCircleLeft, FaArrowCircleRight } from 'react-icons/fa'
import { MdListAlt } from 'react-icons/md'
import { useStorage } from '../../context/localStorage/localStorage'
import { useModal } from '../../context/modals'
import Card, { ICard } from '../CardAnotation'
import { IFormData } from '../FormAddAnnotation'
import * as S from './styles'

export default function Anotations() {
  const [index, setIndex] = useState<number>(0)
  const [cards, setCards] = useState<Array<IFormData>>()
  const [disableNext, setDisableNext] = useState<boolean>(false)
  const [disablePrev, setDisablePrev] = useState<boolean>(true)
  const { setIsVisible, setIsForm } = useModal()
  const { anotations } = useStorage()

  const PERPAGE = 6

  useEffect(() => {
    if (anotations) {
      setCards(anotations)
      if (index + PERPAGE >= anotations.length) {
        setDisableNext(true)
      } else {
        setDisableNext(false)
      }
      if (index - PERPAGE < 0) {
        setDisablePrev(true)
      } else {
        setDisablePrev(false)
      }
    }
  }, [index, anotations])

  const handleNext = () => {
    if (anotations instanceof Array && anotations.length > 0) {
      const newIndex = index + PERPAGE
      setIndex(newIndex)
    }
  }
  const handlePrev = () => {
    if (anotations instanceof Array && anotations.length > 0) {
      const newIndex = index - PERPAGE
      setIndex(newIndex)
    }
  }

  const openForm = () => {
    setIsForm(true)
    setIsVisible(true)
  }

  return (
    <S.Container title="container-cards">
      {cards && cards.length > 0 ? (
        <React.Fragment>
          <S.ContainerCards>
            {cards.slice(index, index + PERPAGE).map((item, i) => (
              <Card key={i} {...(item as ICard)} id={item.id} />
            ))}
          </S.ContainerCards>
          <S.ContainerArrows>
            <button
              disabled={disablePrev}
              onClick={handlePrev}
              aria-label="prev"
            >
              <FaArrowCircleLeft />
            </button>
            <button
              disabled={disableNext}
              onClick={handleNext}
              className="mg"
              aria-label="next"
            >
              <FaArrowCircleRight />
            </button>
          </S.ContainerArrows>
        </React.Fragment>
      ) : (
        <S.ContainerNotFound title="not-found">
          <S.Icon>
            <MdListAlt />
          </S.Icon>
          <S.Paragraph>
            Nenhuma anotação foi encontrada{' '}
            <span onClick={openForm}>Click aqui para adicionar</span>
          </S.Paragraph>
        </S.ContainerNotFound>
      )}
    </S.Container>
  )
}
