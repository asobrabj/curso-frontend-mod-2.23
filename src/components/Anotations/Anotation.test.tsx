import { act, fireEvent, screen, waitFor } from '@testing-library/react'
import Anotations from '.'
import {
  ILocalStorageContext,
  LocalStorageProvider,
  useStorage,
} from '../../context/localStorage/localStorage'
import { IModalContext, ModalProvider, useModal } from '../../context/modals'
import { renderTheme } from '../../utils/styles-test'

jest.mock('../../context/modals')
jest.mock('../../context/localStorage/localStorage')

const mockUseModal = useModal as jest.MockedFunction<typeof useModal>
const mockUseStorage = useStorage as jest.MockedFunction<typeof useStorage>

jest.mock('../../context/localStorage/localStorage', () => {
  const originalModule = jest.requireActual(
    '../../context/localStorage/localStorage'
  )
  return {
    ...originalModule,
    useStorage: jest.fn(() => ({
      anotations: [], // ou os dados desejados para esse teste
    })),
  }
})

jest.mock('../../context/modals', () => {
  const originalModule = jest.requireActual('../../context/modals')
  return {
    ...originalModule,
    useModal: jest.fn(() => ({
      setIsVisible: jest.fn(),
      setIsForm: jest.fn(),
      // Outras funções ou valores que seu componente usa
    })),
  }
})

const mockData = {
  concluded: false,
  date: '2023-12-20',
  name: 'teste',
  time: '00:00',
  category: 'Alimentacao',
  description: 'teste jest',
  priority: 'urgente',
}

const mockModalContext: IModalContext = {
  setIsVisible: jest.fn(),
  setIsForm: jest.fn(),
  setDataCard: jest.fn(),
  setIdCard: jest.fn(),
  isVisible: false,
  isForm: false,
  dataCard: {
    id: null,
    concluded: false,
    date: '',
    description: '',
    name: '',
    category: '',
    priority: '',
    time: '',
  },
  idCard: null
}
const mockLocalStorageContext: ILocalStorageContext = {
  anotations: [
    { id: 1, ...mockData },
    { id: 2, ...mockData },
    { id: 3, ...mockData },
    { id: 4, ...mockData },
    { id: 5, ...mockData },
    { id: 6, ...mockData },
    { id: 7, ...mockData },
  ],
  setAnotations: jest.fn(),
  getFilters: jest.fn(),
  getTags: jest.fn(),
  getStorage: jest.fn(),
  setStorage: jest.fn(),
}

describe('Anotations Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('deve renderizar corretamente quando há anotações', () => {
    mockUseStorage.mockReturnValue(mockLocalStorageContext)

    renderTheme(
      <ModalProvider>
        <LocalStorageProvider>
          <Anotations />
        </LocalStorageProvider>
      </ModalProvider>
    )

    expect(screen.queryByTitle('not-found')).not.toBeInTheDocument()
  })

  it('deve renderizar corretamente quando não há anotações', () => {
    mockUseStorage.mockReturnValue({...mockLocalStorageContext, anotations: []})

    renderTheme(<Anotations />)

    expect(screen.queryByTestId('container-cards')).not.toBeInTheDocument()
    expect(screen.getByTitle('not-found')).toBeInTheDocument()
  })

  it('deve chamar as funções corretas ao clicar nas setas de navegação', () => {
    mockUseStorage.mockReturnValue(mockLocalStorageContext)

    renderTheme(<Anotations />)

    let cards = screen.getAllByTitle('card')
    expect(cards.length).toBe(6)
    act(() => {
      fireEvent.click(screen.getByLabelText('next'))
    })
    cards = screen.getAllByTitle('card')
    expect(cards.length).not.toBe(6)
    expect(cards.length).toBe(1)

    fireEvent.click(screen.getByLabelText('prev'))

    cards = screen.getAllByTitle('card')
    expect(cards.length).not.toBe(1)
    expect(cards.length).toBe(6)
  })

  it('editar card', async () => {
    mockUseStorage.mockReturnValue(mockLocalStorageContext)
    mockUseModal.mockReturnValue(mockModalContext)

    renderTheme(<Anotations />)
    const btn = screen.getAllByText('Editar')[0]

    await act(async () => {
      fireEvent.click(btn)
    })
    waitFor(() => {
      expect(mockModalContext.setIsVisible).toHaveBeenCalledWith(true)
      expect(mockModalContext.setIsForm).toHaveBeenCalledWith(true)
      expect(mockLocalStorageContext.setAnotations).toHaveBeenCalledWith(
        expect.any(Function)
      )
    })
  })

  it('card detail', async () => {
    mockUseStorage.mockReturnValue(mockLocalStorageContext)
    mockUseModal.mockReturnValue(mockModalContext)

    renderTheme(<Anotations />)
    const btn = screen.getAllByTitle('paragraph-name')[0]

    await act(async () => {
      fireEvent.click(btn)
    })
    waitFor(() => {
      expect(mockModalContext.setIsVisible).toHaveBeenCalledWith(true)
      expect(mockModalContext.setIsForm).toHaveBeenCalledWith(true)
      expect(mockLocalStorageContext.setAnotations).toHaveBeenCalledWith(
        expect.any(Function)
      )
    })
  })

  it('delet', async () => {
    mockUseStorage.mockReturnValue(mockLocalStorageContext)
    mockUseModal.mockReturnValue(mockModalContext)

    renderTheme(<Anotations />)
    const btn = screen.getAllByTitle('button delete')[0]

    await act(async () => {
      fireEvent.click(btn)
    })
    waitFor(() => {
      expect(mockModalContext.setIsVisible).toHaveBeenCalledWith(true)
      expect(mockModalContext.setIsForm).toHaveBeenCalledWith(true)
      expect(mockLocalStorageContext.setAnotations).toHaveBeenCalledWith(
        expect.any(Function)
      )
    })
  })

  it('Adicionar', async () => {
    mockUseStorage.mockReturnValue({...mockLocalStorageContext, anotations: []})
    mockUseModal.mockReturnValue(mockModalContext)

    renderTheme(<Anotations />)
    const btn = screen.getByText('Click aqui para adicionar')
    await act(async () => {
      fireEvent.click(btn)
    })
    waitFor(() => {
      expect(mockModalContext.setIsVisible).toHaveBeenCalledWith(true)
      expect(mockModalContext.setIsForm).toHaveBeenCalledWith(true)
      expect(mockLocalStorageContext.setAnotations).toHaveBeenCalledWith(
        expect.any(Function)
      )
    })
  })

  it('snapshot', () => {
    mockUseStorage.mockReturnValue({...mockLocalStorageContext})

    const { container } = renderTheme(<Anotations />)
    expect(container).toMatchSnapshot()
  })
})
