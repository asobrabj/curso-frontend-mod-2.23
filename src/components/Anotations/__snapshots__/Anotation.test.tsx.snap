// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anotations Component snapshot 1`] = `
.c0 {
  padding: 32px;
  margin: auto;
  z-index: 0;
}

.c1 {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.c2 {
  font-size: 150px;
  color: #999;
}

.c3 {
  font-size: 32px;
}

.c3 >span {
  font-style: italic;
}

.c3 >span:hover {
  cursor: pointer;
  text-decoration: underline;
}

<div>
   
  <main
    class="c0"
    title="container-cards"
  >
    <div
      class="c1"
      title="not-found"
    >
      <i
        class="c2"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M19 5v14H5V5h14m1.1-2H3.9c-.5 0-.9.4-.9.9v16.2c0 .4.4.9.9.9h16.2c.4 0 .9-.5.9-.9V3.9c0-.5-.5-.9-.9-.9zM11 7h6v2h-6V7zm0 4h6v2h-6v-2zm0 4h6v2h-6zM7 7h2v2H7zm0 4h2v2H7zm0 4h2v2H7z"
          />
        </svg>
      </i>
      <p
        class="c3"
      >
        Nenhuma anotação foi encontrada
         
        <span>
          Click aqui para adicionar
        </span>
      </p>
    </div>
  </main>
   
</div>
`;
